instant-xml/src/parse.rs

```diff
@@ -1,48 +1,154 @@
-use xmlparser::{ElementEnd, Token};
-
-use super::Error;
-
-impl<'a> Parse for Option<Result<xmlparser::Token<'a>, xmlparser::Error>> {
-    fn element_start(self, ns: Option<&str>, tag: &str) -> Result<(), Error> {
-        match self {
-            Some(Ok(Token::ElementStart { prefix, local, .. })) => {
-                let prefix_ns = prefix.as_str();
-                let (has_prefix, expect_prefix) = (!prefix_ns.is_empty(), ns.is_some());
-                if has_prefix != expect_prefix {
-                    return dbg!(Err(Error::UnexpectedValue));
-                }
+use std::collections::HashMap;
+use std::iter::Peekable;
 
-                if has_prefix && Some(prefix_ns) != ns {
-                    return dbg!(Err(Error::UnexpectedValue));
-                }
+use xmlparser::{ElementEnd, Token, Tokenizer};
+
+use crate::{Error, Result};
+pub use crate::{TagData, XmlRecord};
 
-                if local.as_str() != tag {
-                    return dbg!(Err(Error::UnexpectedValue));
+pub struct XmlParser<'a> {
+    stack: Vec<String>,
+    internal_iter: Peekable<Tokenizer<'a>>,
+}
+
+impl<'a> XmlParser<'a> {
+    pub fn new(input: &'a str) -> XmlParser<'a> {
+        XmlParser {
+            stack: Vec::new(),
+            internal_iter: Tokenizer::from(input).peekable(),
+        }
+    }
+
+    fn parse_next(&mut self) -> Result<Option<XmlRecord>> {
+        let mut key = String::new();
+        let mut prefix_ret = None;
+        let mut default_namespace = None;
+        let mut namespaces = HashMap::new();
+        let mut attributes = Vec::new();
+
+        loop {
+            let item = match self.internal_iter.next() {
+                Some(v) => v,
+                None => return Ok(None),
+            };
+
+            println!("{:?}", &item);
+            match item {
+                Ok(Token::ElementStart { prefix, local, .. }) => {
+                    key = local.to_string();
+                    prefix_ret = if prefix.is_empty() {
```

Comment:
I prefer to write these as `match` expressions, which I think are easier to read:

```rust
match prefix.is_empty() {
    true => None,
    false => Some(prefix.to_string()),
}
```